pymc_marketing/customer_choice/mnl_logit.py

```diff
@@ -0,0 +1,586 @@
+
+        alts = []
+        alt_covariates = []
+        fixed_covariates = []
+        for f in utility_formulas:
+            target, alt_covar, fixed_covar = self.parse_formula(df, f, depvar)
+            f = "0 + " + alt_covar
+            alt_covariates.append(np.asarray(patsy.dmatrix(f, df)).T)
+            alts.append(target)
+            if fixed_covar:
+                fixed_covariates.append(fixed_covar)
+
+        if fixed_covariates:
+            F = np.unique(fixed_covariates)[0]
+            F = "0 + " + F
+            F = np.asarray(patsy.dmatrix(F, df))
+        else:
+            F = []
+
+        X = np.stack(alt_covariates, axis=1).T
+        if X.shape != (n_obs, n_alts, n_covariates):
+            raise ValueError(
+                f"X has shape {X.shape}, but expected {(n_obs, n_alts, n_covariates)}."
+            )
+        return X, F, alts, np.unique(fixed_covariates)
+
+    @staticmethod
+    def _prepare_y_outcome(df, alternatives, depvar):
+        """Encode the outcome category variable for use in the modelling.
+
+        The order of the alterntives should map to the order of the
+        utility formulas.
+        """
+        mode_mapping = dict(zip(alternatives, range(len(alternatives)), strict=False))
+        df["mode_encoded"] = df[depvar].map(mode_mapping)
+        y = df["mode_encoded"].values
+        return y
+
+    @staticmethod
+    def _prepare_coords(df, alternatives, covariates, f_covariates):
+        """Prepare coordinates for PyMC model."""
+        if isinstance(f_covariates, np.ndarray) & (f_covariates.size > 0):
+            f_cov = [s.strip() for s in f_covariates[0].split("+")]
+        else:
+            f_cov = []
+        coords = {
+            "alts": alternatives,
+            "alts_probs": alternatives[:-1],
+            "alt_covariates": covariates,
+            "fixed_covariates": f_cov,
+            "obs": range(len(df)),
+        }
+        return coords
+
+    def preprocess_model_data(self, choice_df, utility_equations):
+        """Pre-process the model initiation inputs into a format that can be used by the PyMC model."""
+        X, F, alternatives, fixed_covar = self.prepare_X_matrix(
+            choice_df, utility_equations, self.depvar
+        )
+        self.X, self.F, self.alternatives, self.fixed_covar = (
+            X,
+            F,
+            alternatives,
+            fixed_covar,
+        )
+        y = self._prepare_y_outcome(choice_df, self.alternatives, self.depvar)
+        self.y = y
+
+        # note: type hints for coords required for mypy to not get confused
+        self.coords: dict[str, list[str]] = self._prepare_coords(
+            choice_df, self.alternatives, self.covariates, self.fixed_covar
+        )
+
+        return X, F, y
+
+    def build_model(self, X, y, **kwargs):
+        """Do not use, required by parent class. Prefer make_model()."""
+        return super().build_model(X, y, **kwargs)
+
+    def make_model(self, X, F, y) -> None:
```

Comment:
Is it possible to put X and F in a single xr.Dataset and pass as "X"

Or would that be unituitive?

Reply:
I think that wouldn't be in the spirit of the model, the decomposition to fixed covariates and alternative specific covariates is pretty foundational to the literature and write up. Would avoid bundling them if i could.